benchmarks/lockhammer/src/lockhammer.c

```diff
@@ -0,0 +1,239 @@
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+ * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+ * IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#define _GNU_SOURCE
+#include <sched.h>
+#include <errno.h>
+#include <unistd.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <pthread.h>
+#include <sys/types.h>
+#include <time.h>
+#include <fcntl.h>
+
+#include ATOMIC_TEST
+
+unsigned long test_lock = 0;
+unsigned long sync_lock = 0;
+unsigned long ready_lock = 0;
+
+struct arg {
+    unsigned long ncores;
+    unsigned long nthrds;
+    unsigned long iter;
+    unsigned long *lock;
+    unsigned long *rst;
+    unsigned long *nsec;
+    unsigned long *depth;
+    unsigned long *nstart;
+    unsigned long hold, post;
+};
+typedef struct arg arg;
+
+void* hmr(void *);
+
+int main(int argc, char** argv)
+{
+    struct sched_param sparam;
+
+    unsigned long i;
+    unsigned long num_cores, num_threads;
+    unsigned long locks_per_thread;
+    unsigned long lock_hold_work, non_lock_work;
+    unsigned long result;
+    unsigned long sched_elapsed = 0, real_elapsed = 0;
+    unsigned long start_ns = 0;
+    double avg_lock_depth = 0.0;
+
+    num_cores = sysconf(_SC_NPROCESSORS_ONLN);
+
+    if (argc == 1) {
+        num_threads = num_cores;
+        locks_per_thread = 50000;
+        lock_hold_work = 0;
+        non_lock_work = 0;
+    }
+    else if (argc == 5) {
+        num_threads = atoi(argv[1]);
+	/* Do not allow number of threads to exceed online cores
+	   in order to prevent deadlock ... */
+        num_threads = num_threads > num_cores ? num_cores : num_threads;
+        locks_per_thread = atoi(argv[2]);
+        lock_hold_work = atoi(argv[3]);
+        non_lock_work = atoi(argv[4]);
+    }
+    else {
+        fprintf(stderr, "Usage: %s [<cores> <threads per core> <critical loops> <post-release loops>]\n", argv[0]);
+        return 1;
+    }
+
+    pthread_t hmr_threads[num_threads];
+    pthread_attr_t hmr_attr;
+    unsigned long hmrs[num_threads];
+    unsigned long hmrtime[num_threads]; /* can't touch this */
+    unsigned long hmrdepth[num_threads];
+    struct timespec tv_time;
+
+    for (i = 0; i < num_threads; ++i) hmrs[i] = 0;
+
+    pthread_attr_init(&hmr_attr);
+    pthread_attr_setinheritsched(&hmr_attr, PTHREAD_EXPLICIT_SCHED);
+    pthread_attr_setschedpolicy(&hmr_attr, SCHED_FIFO);
+    sparam.sched_priority = 1;
+    pthread_attr_setschedparam(&hmr_attr, &sparam);
```

Comment:
This block of code actually ends up causing our machines to throw kernel panics from unacknowledged I/O due to how we have them setup and for some, they eventually become unresponsive and simply crash, requiring a power-cycle.  How much variation does having this in here eliminate?  Can we make this a command line flag option at the very least and document it as a potentially dangerous option?

Reply:
Does reverting to the shorter 50k (or smaller) acquires per thread mitigate this?  For non-mutex synchronization algorithms like the refcounts the impact of not running in the FIFO scheduler is quite small.  But for mutexes, especially semi-fair ones like the kernel spinlocks the impact of a single thread getting rescheduled can be massive (several times the normal lock acquisition time) since the lock may be held or the queue blocked in the interim.  Obviously for userspace mutexes this can be a real problem but even in those cases I feel that we want to ignore those impacts to be able to compare algorithms and SUTs on as level a playing field as possible. 
In general, our experiences are that runing with FIFO enabled at even a couple of thousand acquires per thread gives more consistent results than hundreds of thousands of acquires per thread in the normal scheduler.  I'm not against making this an option but it does make a big difference in the usefulness of the tool for certain types of locks.

Reply:
I see your point.  We'll test with the smaller runtime and see if that helps.  We'll internally test and also see if moving IRQs around helps too.  Adding some documentation around this will be a minimum though as we hit the issues and it was not immediately obvious as to why the system was becoming unresponsive.

Reply:
Ok, please update me with your findings.  For now I will roll back the change to 200k acquires/thread in Zoy's diff (and probably select an even lower default for safety's sake) and push the other requested changes.  We can revisit alternative default runtimes then.

Reply:
Ok, thanks for this Lucas.  I'm waiting on Zoy to get back with some test results from our suite of machines to see if we can keep the # of acquires input as a runtime proxy.  We may have to move to a timer based method if the current method is too unstable since trying to automate these tests like we want to do will become difficult if we have to tune per machine, per core count, per lock.

Reply:
We tried 50k on all our machines, only 1 machine would always generate kernel timeout oops.
For that specific machine, we tried several loop iterations (200k, 50k, 30k, 10k, 5k, 3k, 1k), only 1k passed the swap-mutex 1000 5000 sweep test.
If we use 200k iterations, all machines would kernel panic at different thread count. Nobody can sweep to the maximum cores.

Reply:
I say we document this thoroughly for now.  This is something we will have to fix via another issue in the tracker.

Reply:
Opened new issue https://github.com/ARM-software/synchronization-benchmarks/issues/3
Because only 1 machine has this problem, we would keep the default value 50000.

Reply:
I will add another commit to the PR to return the default to 50000.  Zoy, can you try the empty test on a single core with 1000 critical time and 0 non-critical time and just a few loops (eg: sudo ./build/lh_empty 1 100 1000 0) on both the affected and some non-affecetd systems?  If the inter-acquire time is significantly larger on the affected system than other systems then that may be exacerbating the problem and a different value may be necessary.  In general these iteration counts (1000 and 5000) need to be tuned on a per-system basis in the current code but in practice they have proven to be pretty close from system to system on the things we've tested.  If you can confirm that this is the cause of the issue then I will work on adding some benchmarking of the busy loop to tune the actual iteration counts prior to test start based on individual system performance and convert the command line args to number of nanoseconds or something else more generic instead of busy loop iteration counts.

Reply:
Do you mean the access time per each iteration? I did the following tests as you mentioned:

For lh_empty (./lh_empty 1 50000 1000 0): problematic machine uses about 500ns per access, the good machines' (all arm machines) value range from 900ns to 1500ns.
For lh_swap_mutex (./lh_swap_mutex 1 50000 1000 0), per access time only increased 4~7% respectively.

The problematic machine has higher per core hyperthread count than Xeon, probably that is the reason why high core count synchronization performance is not good.

Reply:
Oh, I forgot to mention this.  At least on Skylake we see 0 scaling in the "busy" loops across hardware threads.  There is no real work done so probably the busy loop is fetch or retire bandwidth limited and so even a single hardware thread can saturate the physical core's resources.  When subsequent threads are added the number of iterations on that core increases but the throughput does not.  So on a hyperthreaded system you may need to cut these iterations (1000, 5000) by the number of threads per core (the empty test can be helpful for this by testing 1 core from 1 to max number of hardware threads).  Otherwise the critical section length will grow based on the number of hardware threads rather than staying constant.

Reply:
We tested the new dynamic per thread lock acquire count scheme (maximum 400k total threads, minimum 1k per thread), and the test result is 3~5% different from fixed 200k per thread lock acquire count. The only issue is we found a performance regression around 32 exectx, probably because of interconnect or something. So would you mind increase the single step phase from 24 to 48? e.g. `seq 48`. Seems 24 is not enough for some cases. Thanks.

Reply:
Hi Zoy, I added the requested change in the latest commit.  Please check 2e2b072 to confirm.

Reply:
Great, since there is no other issue for all our machines, we will checkin and close this pull request this afternoon.